Problem with Dotfuscator

I couldn't compile my project using Dotfuscator:

> Symbols were found but are not matching the assembly
   at Mono.Cecil.ModuleDefinition.ReadSymbols(ISymbolReader reader, Boolean throwIfSymbolsAreNotMaching)
   at Mono.Cecil.ModuleReader.ReadSymbols(ModuleDefinition module, ReaderParameters parameters)
   at Mono.Cecil.ModuleReader.CreateModule(Image image, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(String fileName, ReaderParameters parameters)
   at Xamarin.AndroidX.Migration.CecilMigrator.Migrate(String source, String destination) in D:\a\1\s\source\Xamarin.AndroidX.Migration\CecilMigrator.cs:line 85
   at Xamarin.AndroidX.Migration.CecilMigrator.Migrate(IEnumerable`1 assemblies) in D:\a\1\s\source\Xamarin.AndroidX.Migration\CecilMigrator.cs:line 50
   at Xamarin.AndroidX.Migration.BuildTasks.CecilfyFiles.Execute() in D:\a\1\s\source\Xamarin.AndroidX.Migration.BuildTasks\CecilfyFiles.cs:line 35	x2Apps.CheckPlaca.App.Android	C:\Users\mario\.nuget\packages\xamarin.androidx.migration\1.0.0-preview02\build\Xamarin.AndroidX.Migration.targets	182	

Setting the DotfuscatorXamarinEnabled option to FALSE in my csproj solves the problem.

Thanks @tranb3r.

Did anyone found a solution for Dotfuscator (Community)?
@MarvinFrankenfeld solution hasn't worked for me.

I've tried to add "_AndroidXCecilfyShrink" to the build entry point of **PreEmptive.Dotfuscator.Xamarin.Targets**:

```
   <!--                    -->
   <!-- Build entry points -->
   <!--                    -->
 
   <Target Name="DotfuscatorXamarin_Build"
           DependsOnTargets="DotfuscatorXamarin_Validate;
                             DotfuscatorXamarin_Prepare;
                             DotfuscatorXamarin_Protect;
                             DotfuscatorXamarin_Update;
                             _AndroidXCecilfyShrink"/>
```

But this gives me another error:
```
{Path}\Xamarin.AndroidX.Migration.targets(227,9): error MSB4018: Unexpected error CecilfyFiles-task. [{MyProjectPath}.csproj]
{Path}\Xamarin.AndroidX.Migration.targets(227,9): error MSB4018: System.ArgumentNullException: Value cannot be null. [{MyProjectPath}.csproj]
{Path}\Xamarin.AndroidX.Migration.targets(227,9): error MSB4018: Parametername: source [{MyProjectPath}.csproj]
{Path}\Xamarin.AndroidX.Migration.targets(227,9): error MSB4018:    bei System.Linq.Enumerable.Select[TSource,TResult](IEnumerable`1 source, Func`2 selector) [{MyProjectPath}.csproj]
{Path}\Xamarin.AndroidX.Migration.targets(227,9): error MSB4018:    bei Xamarin.AndroidX.Migration.BuildTasks.CecilfyFiles.Execute() [{MyProjectPath}.csproj]
{Path}\Xamarin.AndroidX.Migration.targets(227,9): error MSB4018:    bei Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [{MyProjectPath}.csproj]
{Path}\Xamarin.AndroidX.Migration.targets(227,9): error MSB4018:    bei Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [{MyProjectPath}.csproj]
```

--- a/source/migration/Migration/CecilMigrator.cs
+++ b/source/migration/Migration/CecilMigrator.cs
@@ -1,6 +1,5 @@
 ﻿using Mono.Cecil;
 using Mono.Cecil.Cil;
-using Mono.Cecil.Pdb;
 using System;
 using System.Collections.Generic;
 using System.IO;
@@ -97,19 +96,9 @@ namespace Xamarin.AndroidX.Migration
 					}
 				}
 
-				var readerParams = new ReaderParameters
-				{
-					ReadSymbols = hasSymbols,
-					ReadWrite = true,
-					AssemblyResolver = resolver,
-				};
-
 				LogVerboseMessage($"Processing assembly '{source}'...");
-				using (var assembly = AssemblyDefinition.ReadAssembly(destination, readerParams))
+				using (var assembly = ReadAssembly(resolver, destination))
 				{
-					if (!hasSymbols)
-						LogVerboseMessage($"  No debug symbols found for the assembly.");
-
 					result = MigrateAssembly(assembly);
 
 					requiresSave =
@@ -580,6 +569,41 @@ namespace Xamarin.AndroidX.Migration
 			File.Copy(src, dst, true);
 
 			return true;
+		}
+
+		private AssemblyDefinition ReadAssembly(AssemblyResolver resolver, string path)
+		{
+			var readerParams = new ReaderParameters
+			{
+				ReadSymbols = true,
+				ReadWrite = true,
+				AssemblyResolver = resolver,
+			};
+
+			var pdbPath = Path.ChangeExtension(path, "pdb");
+			var mdbPath = path + ".mdb";
+
+			if (File.Exists(pdbPath) || File.Exists(mdbPath))
+			{
+				// if there are symbols, then try using them
+				try
+				{
+					return AssemblyDefinition.ReadAssembly(path, readerParams);
+				}
+				catch (Exception ex)
+				{
+					// do nothing
+					LogMessage($"  Debug symbols were found for the assembly, but they couldn't be used: " + ex.Message);
+				}
+			}
+			else
+			{
+				LogVerboseMessage($"  No debug symbols found for the assembly.");
+			}
+
+			// just ignore all symbols, there is nothing that can be done about bad symbols
+			readerParams.ReadSymbols = false;
+			return AssemblyDefinition.ReadAssembly(path, readerParams);
 		}
 
 		private sealed class AssemblyResolver : DefaultAssemblyResolver

--- a/tests/AndroidXMigrationTests/Tests/BaseTests.cs
+++ b/tests/AndroidXMigrationTests/Tests/BaseTests.cs
@@ -12,6 +12,7 @@ namespace Xamarin.AndroidX.Migration.Tests
 		public const string MergedSupportDll = "AndroidSupport.Merged.dll";
 		public const string MergedAndroidXDll = "AndroidX.Merged.dll";
 		public const string ActiveDirectoryDll = "Microsoft.IdentityModel.Clients.ActiveDirectory.dll";
+		public const string ActiveDirectoryPdb = "Microsoft.IdentityModel.Clients.ActiveDirectory.pdb";
 
 		public const string ManagedSupportDll = "Aarxercise.Managed.Support.dll";
 		public const string ManagedAndroidXDll = "Aarxercise.Managed.AndroidX.dll";
@@ -90,11 +91,26 @@ namespace Xamarin.AndroidX.Migration.Tests
 			return destFile;
 		}
 
-		public static string RunMigration(string supportDll, CecilMigrationResult expectedResult)
+		public static string RunMigration(string supportDll, CecilMigrationResult expectedResult, string logPath = null)
 		{
 			var migratedDll = Utils.GetTempFilename() + ".dll";
 
 			var migrator = new CecilMigrator();
+
+			if (!string.IsNullOrWhiteSpace(logPath))
+			{
+				migrator.MessageLogged += (sender, e) =>
+				{
+					var log = string.Empty;
+					if (e.IsError)
+						log += "ERROR: ";
+					if (e.IsVerbose)
+						log += "VERBOSE: ";
+					log += e.Message;
+					File.AppendAllLines(logPath, new [] { log });
+				};
+			}
+
 			var result = migrator.Migrate(supportDll, migratedDll);
 
 			Assert.Equal(expectedResult, result);

--- a/tests/AndroidXMigrationTests/Tests/BaseTests.cs
+++ b/tests/AndroidXMigrationTests/Tests/BaseTests.cs
@@ -1,6 +1,8 @@
-﻿using System.IO;
+﻿using System.Collections.Generic;
+using System.IO;
 using System.IO.Compression;
 using System.Linq;
+using Mono.Cecil;
 using Xunit;
 
 namespace Xamarin.AndroidX.Migration.Tests
@@ -124,6 +126,18 @@ namespace Xamarin.AndroidX.Migration.Tests
 			using var br = new BinaryReader(fs);
 			
 			return br.ReadUInt32() == PortablePdbSignature;
+		}
+
+		public static void AssertHasSupportTypes(IEnumerable<TypeReference> types)
+		{
+			Assert.NotEmpty(types.Where(t => t.FullName.StartsWith("Android.Support.") || t.FullName.StartsWith("Android.Arch.")));
+		}
+
+		public static void AssertNoSupportTypes(IEnumerable<TypeReference> types)
+		{
+			Assert.Empty(types.Where(t => 
+				(t.FullName.StartsWith("Android.Support.") || t.FullName.StartsWith("Android.Arch.")) &&
+				(t.FullName != "Android.Support.V4.Media.Session.MediaSessionCompat")));
 		}
 	}
 }

--- a/tests/AndroidXMigrationTests/Tests/ManagedMigrationTests.cs
+++ b/tests/AndroidXMigrationTests/Tests/ManagedMigrationTests.cs
@@ -12,6 +12,7 @@ namespace Xamarin.AndroidX.Migration.Tests
 		[InlineData(ManagedSupportDll, true)]
 		[InlineData(OldSupportDll, true)]
 		[InlineData(BindingSupportDll, true)]
+		[InlineData(ActiveDirectoryDll, true)]
 		[InlineData(MergedSupportDll, false)]
 		public void AssembliesHaveSupportReferences(string assembly, bool hasSupportReference)
 		{
@@ -142,6 +143,29 @@ namespace Xamarin.AndroidX.Migration.Tests
 						XDocument.Load(migratedLayout).Root.Elements().FirstOrDefault().Name.LocalName);
 				}
 			}
+		}
+
+		[Theory]
+		[InlineData(ActiveDirectoryDll, ActiveDirectoryPdb, CecilMigrationResult.ContainedSupport | CecilMigrationResult.PotentialJavaArtifacts)]
+		[InlineData(ManagedSupportDll, ActiveDirectoryPdb, CecilMigrationResult.ContainedSupport | CecilMigrationResult.PotentialJavaArtifacts | CecilMigrationResult.ContainedJavaArtifacts)]
+		[InlineData(ManagedSupportDll, ManagedSupportDll, CecilMigrationResult.ContainedSupport | CecilMigrationResult.PotentialJavaArtifacts | CecilMigrationResult.ContainedJavaArtifacts)]
+		public void CanRunMigrationOnCorruptedPdbs(string assembly, string symbols, CecilMigrationResult expectedResult)
+		{
+			var dll = Utils.GetTempFilename(Path.GetFileName(assembly));
+			var pdb = Path.ChangeExtension(dll, "pdb");
+			File.Copy(assembly, dll, true);
+			File.Copy(symbols, pdb, true);
+			
+			var migratedDll = RunMigration(dll, expectedResult);
+
+			using (var def = AssemblyDefinition.ReadAssembly(migratedDll))
+			{
+				var types = def.MainModule.GetTypeReferences();
+				Assert.Empty(types.Where(t => 
+					(t.FullName.StartsWith("Android.Support.") || t.FullName.StartsWith("Android.Arch.")) &&
+					(t.FullName != "Android.Support.V4.Media.Session.MediaSessionCompat")));
+			}
+			Assert.True(IsPortablePdb(Path.ChangeExtension(migratedDll, "pdb")), migratedDll);
 		}
 	}
 }

--- a/tests/AndroidXMigrationTests/Tests/ManagedMigrationTests.cs
+++ b/tests/AndroidXMigrationTests/Tests/ManagedMigrationTests.cs
@@ -19,11 +19,10 @@ namespace Xamarin.AndroidX.Migration.Tests
 			using (var support = AssemblyDefinition.ReadAssembly(assembly))
 			{
 				var types = support.MainModule.GetTypeReferences();
-				var supportReferences = types.Where(t => t.FullName.StartsWith("Android.Support.") || t.FullName.StartsWith("Android.Arch."));
 				if (hasSupportReference)
-					Assert.NotEmpty(supportReferences);
+					AssertHasSupportTypes(types);
 				else
-					Assert.Empty(supportReferences);
+					AssertNoSupportTypes(types);
 			}
 		}
 
@@ -161,9 +160,7 @@ namespace Xamarin.AndroidX.Migration.Tests
 			using (var def = AssemblyDefinition.ReadAssembly(migratedDll))
 			{
 				var types = def.MainModule.GetTypeReferences();
-				Assert.Empty(types.Where(t => 
-					(t.FullName.StartsWith("Android.Support.") || t.FullName.StartsWith("Android.Arch.")) &&
-					(t.FullName != "Android.Support.V4.Media.Session.MediaSessionCompat")));
+				AssertNoSupportTypes(types);
 			}
 			Assert.True(IsPortablePdb(Path.ChangeExtension(migratedDll, "pdb")), migratedDll);
 		}

